File: src/main/java/com/exsoloscript/challonge/model/Tournament.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 EXSolo <exsoloscripter at gmail.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.exsoloscript.challonge.model;

import com.exsoloscript.challonge.model.enumeration.RankedBy;
import com.exsoloscript.challonge.model.enumeration.TournamentState;
import com.exsoloscript.challonge.model.enumeration.TournamentType;
import com.exsoloscript.challonge.model.enumeration.query.GrandFinalsModifier;
import com.google.gson.annotations.SerializedName;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.experimental.Accessors;

import java.time.OffsetDateTime;
import java.util.List;

/**
 * The POJO that will be mapped to the tournament requests by Gson
 *
 * @author EXSolo
 * @version 20160820.1
 */
@Data
@Accessors(fluent = true)
@EqualsAndHashCode(exclude = {
        "startAt", "completedAt", "createdAt",
        "predictionsOpenedAt", "startedAt", "startedCheckingInAt",
        "updatedAt", "lockedAt"})
public class Tournament {
    private final String name;
    @SerializedName("tournament_type")
    private final TournamentType tournamentType;
    private final String url;
    private final String subdomain;
    private final String description;
    @SerializedName("open_signup")
    private final Boolean openSignup;
    @SerializedName("hold_third_place_match")
    private final Boolean holdThirdPlaceMatch;
    @SerializedName("pts_for_match_win")
    private final Float pointsForMatchWin;
    @SerializedName("pts_for_match_tie")
    private final Float pointsForMatchTie;
    @SerializedName("pts_for_game_win")
    private final Float pointsForGameWin;
    @SerializedName("pts_for_game_tie")
    private final Float pointsForGameTie;
    @SerializedName("pts_for_bye")
    private final Float pointsForBye;
    @SerializedName("swiss_rounds")
    private final Integer swissRounds;
    @SerializedName("ranked_by")
    private final RankedBy rankedBy;
    @SerializedName("rr_pts_for_game_win")
    private final Float roundRobinPointsForGameWin;
    @SerializedName("rr_pts_for_game_tie")
    private final Float roundRobinPointsForGameTie;
    @SerializedName("rr_pts_for_match_win")
    private final Float roundRobinPointsForMatchWin;
    @SerializedName("rr_pts_for_match_tie")
    private final Float roundRobinPointsForMatchTie;
    @SerializedName("accept_attachments")
    private final Boolean acceptAttachments;
    @SerializedName("hide_forum")
    private final Boolean hideForum;
    @SerializedName("show_rounds")
    private final Boolean showRounds;
    @SerializedName("private")
    private final Boolean _private;
    @SerializedName("notify_users_when_the_tournament_ends")
    private final Boolean notifyUsersWhenTheTournamentEnds;
    @SerializedName("sequential_pairings")
    private final Boolean sequentialPairings;
    @SerializedName("signup_cap")
    private final Integer signupCap;
    @SerializedName("start_at")
    private final OffsetDateTime startAt;
    @SerializedName("check_in_duration")
    private final Integer checkInDuration;
    @SerializedName("allow_participant_match_reporting")
    private final Boolean allowParticipantMatchReporting;
    @SerializedName("anonymous_voting")
    private final Boolean anonymousVoting;
    private final String category;
    @SerializedName("completed_at")
    private final OffsetDateTime completedAt;
    @SerializedName("created_at")
    private final OffsetDateTime createdAt;
    @SerializedName("created_by_api")
    private final Boolean createdByApi;
    @SerializedName("credit_capped")
    private final Boolean creditCapped;
    @SerializedName("game_id")
    private final Integer gameId;
    @SerializedName("group_stages_enabled")
    private final Boolean groupStagesEnabled;
    @SerializedName("hide_seeds")
    private final Boolean hideSeeds;
    private final Integer id;
    @SerializedName("max_predictions_per_user")
    private final Integer maxPredictionsPerUser;
    @SerializedName("notify_users_when_matches_open")
    private final Boolean notifyUsersWhenMatchesOpen;
    @SerializedName("participants_count")
    private final Integer participantsCount;
    @SerializedName("prediction_method")
    private final Integer predictionMethod;
    @SerializedName("predictions_opened_at")
    private final OffsetDateTime predictionsOpenedAt;
    @SerializedName("progress_meter")
    private final Integer progressMeter;
    @SerializedName("quick_advance")
    private final Boolean quickAdvance;
    @SerializedName("require_score_agreement")
    private final Boolean requireScoreAgreement;
    @SerializedName("started_at")
    private final OffsetDateTime startedAt;
    @SerializedName("started_checking_in_at")
    private final OffsetDateTime startedCheckingInAt;
    private final TournamentState state;
    private final Boolean teams;
    @SerializedName("tie_breaks")
    private final String[] tieBreaks;
    @SerializedName("updated_at")
    private final OffsetDateTime updatedAt;
    @SerializedName("description_source")
    private final String descriptionSource;
    @SerializedName("full_challonge_url")
    private final String fullChallongeUrl;
    @SerializedName("live_image_url")
    private final String liveImageUrl;
    @SerializedName("sign_up_url")
    private final String signUpUrl;
    @SerializedName("review_before_finalizing")
    private final Boolean reviewBeforeFinalizing;
    @SerializedName("accepting_predictions")
    private final Boolean acceptingPredictions;
    @SerializedName("participants_locked")
    private final Boolean participantsLocked;
    @SerializedName("game_name")
    private final String gameName;
    @SerializedName("participants_swappable")
    private final Boolean participantsSwappable;
    @SerializedName("team_convertable")
    private final Boolean teamConvertable;
    @SerializedName("group_stages_were_started")
    private final Boolean groupStagesWereStarted;
    @SerializedName("locked_at")
    private final OffsetDateTime lockedAt;
    @SerializedName("event_id")
    private final Integer eventId;
    @SerializedName("public_predictions_before_start_time")
    private final Boolean publicPredictionsBeforeStartTime;
    @SerializedName("grand_finals_modifier")
    private final GrandFinalsModifier grandFinalsModifier;
    private final List<Participant> participants;
    private final List<Match> matches;
}

File: core/src/main/kotlin/at/stefangeyer/challonge/model/Tournament.kt
/*
 * Copyright 2017 Stefan Geyer <stefangeyer at outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package at.stefangeyer.challonge.model

import at.stefangeyer.challonge.model.enum.RankedBy
import at.stefangeyer.challonge.model.enum.TournamentState
import at.stefangeyer.challonge.model.enum.TournamentType
import at.stefangeyer.challonge.model.query.enum.GrandFinalsModifier
import java.time.OffsetDateTime

class Tournament(
        var id: Long = 0L,
        var url: String = "",
        var name: String? = null,
        var tournamentType: TournamentType = TournamentType.SINGLE_ELIMINATION,
        var subdomain: String? = null,
        var description: String? = null,
        var openSignup: Boolean = false,
        var holdThirdPlaceMatch: Boolean = false,
        var pointsForMatchWin: Float = 0.0F,
        var pointsForMatchTie: Float = 0.0F,
        var pointsForGameWin: Float = 0.0F,
        var pointsForGameTie: Float = 0.0F,
        var pointsForBye: Float = 0.0F,
        var swissRounds: Int = 0,
        var rankedBy: RankedBy? = null,
        var roundRobinPointsForGameWin: Float = 0.0F,
        var roundRobinPointsForGameTie: Float = 0.0F,
        var roundRobinPointsForMatchWin: Float = 0.0F,
        var roundRobinPointsForMatchTie: Float = 0.0F,
        var acceptAttachments: Boolean = false,
        var hideForum: Boolean = false,
        var showRounds: Boolean = false,
        var private: Boolean = false,
        var notifyUsersWhenTheTournamentEnds: Boolean = false,
        var sequentialPairings: Boolean = false,
        var signupCap: Int? = 0,
        var startAt: OffsetDateTime? = null,
        var checkInDuration: Long? = 0,
        var allowParticipantMatchReporting: Boolean = false,
        var anonymousVoting: Boolean = false,
        var category: String? = null,
        var completedAt: OffsetDateTime? = null,
        var createdAt: OffsetDateTime? = null,
        var createdByApi: Boolean = false,
        var creditCapped: Boolean = false,
        var gameId: Long? = null,
        var groupStagesEnabled: Boolean = false,
        var hideSeeds: Boolean = false,
        var maxPredictionsPerUser: Int = 0,
        var notifyUsersWhenMatchesOpen: Boolean = false,
        var participantsCount: Int = 0,
        var predictionMethod: Int = 0,
        var predictionsOpenedAt: OffsetDateTime? = null,
        var progressMeter: Int = 0,
        var quickAdvance: Boolean = false,
        var requireScoreAgreement: Boolean = false,
        var startedAt: OffsetDateTime? = null,
        var startedCheckingInAt: OffsetDateTime? = null,
        var state: TournamentState = TournamentState.PENDING,
        var teams: Boolean? = false,
        var tieBreaks: List<String>? = listOf(),
        var updatedAt: OffsetDateTime? = null,
        var descriptionSource: String? = null,
        var fullChallongeUrl: String? = null,
        var liveImageUrl: String? = null,
        var signUpUrl: String? = null,
        var reviewBeforeFinalizing: Boolean = false,
        var acceptingPredictions: Boolean = false,
        var participantsLocked: Boolean = false,
        var gameName: String? = null,
        var participantsSwappable: Boolean = false,
        var teamConvertable: Boolean = false,
        var groupStagesWereStarted: Boolean = false,
        var lockedAt: OffsetDateTime? = null,
        var eventId: Long? = 0L,
        var publicPredictionsBeforeStartTime: Boolean? = false,
        var grandFinalsModifier: GrandFinalsModifier? = null,
        var participants: List<Participant> = listOf(),
        var matches: List<Match> = listOf()
) {


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Tournament

        if (id != other.id) return false
        if (url != other.url) return false
        if (name != other.name) return false
        if (tournamentType != other.tournamentType) return false
        if (subdomain != other.subdomain) return false
        if (description != other.description) return false
        if (openSignup != other.openSignup) return false
        if (holdThirdPlaceMatch != other.holdThirdPlaceMatch) return false
        if (pointsForMatchWin != other.pointsForMatchWin) return false
        if (pointsForMatchTie != other.pointsForMatchTie) return false
        if (pointsForGameWin != other.pointsForGameWin) return false
        if (pointsForGameTie != other.pointsForGameTie) return false
        if (pointsForBye != other.pointsForBye) return false
        if (swissRounds != other.swissRounds) return false
        if (rankedBy != other.rankedBy) return false
        if (roundRobinPointsForGameWin != other.roundRobinPointsForGameWin) return false
        if (roundRobinPointsForGameTie != other.roundRobinPointsForGameTie) return false
        if (roundRobinPointsForMatchWin != other.roundRobinPointsForMatchWin) return false
        if (roundRobinPointsForMatchTie != other.roundRobinPointsForMatchTie) return false
        if (acceptAttachments != other.acceptAttachments) return false
        if (hideForum != other.hideForum) return false
        if (showRounds != other.showRounds) return false
        if (private != other.private) return false
        if (notifyUsersWhenTheTournamentEnds != other.notifyUsersWhenTheTournamentEnds) return false
        if (sequentialPairings != other.sequentialPairings) return false
        if (signupCap != other.signupCap) return false
        if (checkInDuration != other.checkInDuration) return false
        if (allowParticipantMatchReporting != other.allowParticipantMatchReporting) return false
        if (anonymousVoting != other.anonymousVoting) return false
        if (category != other.category) return false
        if (createdByApi != other.createdByApi) return false
        if (creditCapped != other.creditCapped) return false
        if (gameId != other.gameId) return false
        if (groupStagesEnabled != other.groupStagesEnabled) return false
        if (hideSeeds != other.hideSeeds) return false
        if (maxPredictionsPerUser != other.maxPredictionsPerUser) return false
        if (notifyUsersWhenMatchesOpen != other.notifyUsersWhenMatchesOpen) return false
        if (participantsCount != other.participantsCount) return false
        if (predictionMethod != other.predictionMethod) return false
        if (progressMeter != other.progressMeter) return false
        if (quickAdvance != other.quickAdvance) return false
        if (requireScoreAgreement != other.requireScoreAgreement) return false
        if (state != other.state) return false
        if (teams != other.teams) return false
        if (tieBreaks != other.tieBreaks) return false
        if (descriptionSource != other.descriptionSource) return false
        if (fullChallongeUrl != other.fullChallongeUrl) return false
        if (liveImageUrl != other.liveImageUrl) return false
        if (signUpUrl != other.signUpUrl) return false
        if (reviewBeforeFinalizing != other.reviewBeforeFinalizing) return false
        if (acceptingPredictions != other.acceptingPredictions) return false
        if (participantsLocked != other.participantsLocked) return false
        if (gameName != other.gameName) return false
        if (participantsSwappable != other.participantsSwappable) return false
        if (teamConvertable != other.teamConvertable) return false
        if (groupStagesWereStarted != other.groupStagesWereStarted) return false
        if (eventId != other.eventId) return false
        if (publicPredictionsBeforeStartTime != other.publicPredictionsBeforeStartTime) return false
        if (grandFinalsModifier != other.grandFinalsModifier) return false
        if (participants != other.participants) return false
        if (matches != other.matches) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + url.hashCode()
        result = 31 * result + (name?.hashCode() ?: 0)
        result = 31 * result + tournamentType.hashCode()
        result = 31 * result + (subdomain?.hashCode() ?: 0)
        result = 31 * result + (description?.hashCode() ?: 0)
        result = 31 * result + openSignup.hashCode()
        result = 31 * result + holdThirdPlaceMatch.hashCode()
        result = 31 * result + pointsForMatchWin.hashCode()
        result = 31 * result + pointsForMatchTie.hashCode()
        result = 31 * result + pointsForGameWin.hashCode()
        result = 31 * result + pointsForGameTie.hashCode()
        result = 31 * result + pointsForBye.hashCode()
        result = 31 * result + swissRounds
        result = 31 * result + (rankedBy?.hashCode() ?: 0)
        result = 31 * result + roundRobinPointsForGameWin.hashCode()
        result = 31 * result + roundRobinPointsForGameTie.hashCode()
        result = 31 * result + roundRobinPointsForMatchWin.hashCode()
        result = 31 * result + roundRobinPointsForMatchTie.hashCode()
        result = 31 * result + acceptAttachments.hashCode()
        result = 31 * result + hideForum.hashCode()
        result = 31 * result + showRounds.hashCode()
        result = 31 * result + private.hashCode()
        result = 31 * result + notifyUsersWhenTheTournamentEnds.hashCode()
        result = 31 * result + sequentialPairings.hashCode()
        result = 31 * result + (signupCap ?: 0)
        result = 31 * result + (checkInDuration?.hashCode() ?: 0)
        result = 31 * result + allowParticipantMatchReporting.hashCode()
        result = 31 * result + anonymousVoting.hashCode()
        result = 31 * result + (category?.hashCode() ?: 0)
        result = 31 * result + createdByApi.hashCode()
        result = 31 * result + creditCapped.hashCode()
        result = 31 * result + (gameId?.hashCode() ?: 0)
        result = 31 * result + groupStagesEnabled.hashCode()
        result = 31 * result + hideSeeds.hashCode()
        result = 31 * result + maxPredictionsPerUser
        result = 31 * result + notifyUsersWhenMatchesOpen.hashCode()
        result = 31 * result + participantsCount
        result = 31 * result + predictionMethod
        result = 31 * result + progressMeter
        result = 31 * result + quickAdvance.hashCode()
        result = 31 * result + requireScoreAgreement.hashCode()
        result = 31 * result + state.hashCode()
        result = 31 * result + (teams?.hashCode() ?: 0)
        result = 31 * result + (tieBreaks?.hashCode() ?: 0)
        result = 31 * result + (descriptionSource?.hashCode() ?: 0)
        result = 31 * result + (fullChallongeUrl?.hashCode() ?: 0)
        result = 31 * result + (liveImageUrl?.hashCode() ?: 0)
        result = 31 * result + (signUpUrl?.hashCode() ?: 0)
        result = 31 * result + reviewBeforeFinalizing.hashCode()
        result = 31 * result + acceptingPredictions.hashCode()
        result = 31 * result + participantsLocked.hashCode()
        result = 31 * result + (gameName?.hashCode() ?: 0)
        result = 31 * result + participantsSwappable.hashCode()
        result = 31 * result + teamConvertable.hashCode()
        result = 31 * result + groupStagesWereStarted.hashCode()
        result = 31 * result + (eventId?.hashCode() ?: 0)
        result = 31 * result + (publicPredictionsBeforeStartTime?.hashCode() ?: 0)
        result = 31 * result + (grandFinalsModifier?.hashCode() ?: 0)
        result = 31 * result + participants.hashCode()
        result = 31 * result + matches.hashCode()
        return result
    }

    override fun toString(): String {
        return "Tournament(id=$id, url='$url', name=$name, tournamentType=$tournamentType, subdomain=$subdomain, description=$description, openSignup=$openSignup, holdThirdPlaceMatch=$holdThirdPlaceMatch, pointsForMatchWin=$pointsForMatchWin, pointsForMatchTie=$pointsForMatchTie, pointsForGameWin=$pointsForGameWin, pointsForGameTie=$pointsForGameTie, pointsForBye=$pointsForBye, swissRounds=$swissRounds, rankedBy=$rankedBy, roundRobinPointsForGameWin=$roundRobinPointsForGameWin, roundRobinPointsForGameTie=$roundRobinPointsForGameTie, roundRobinPointsForMatchWin=$roundRobinPointsForMatchWin, roundRobinPointsForMatchTie=$roundRobinPointsForMatchTie, acceptAttachments=$acceptAttachments, hideForum=$hideForum, showRounds=$showRounds, private=$private, notifyUsersWhenTheTournamentEnds=$notifyUsersWhenTheTournamentEnds, sequentialPairings=$sequentialPairings, signupCap=$signupCap, startAt=$startAt, checkInDuration=$checkInDuration, allowParticipantMatchReporting=$allowParticipantMatchReporting, anonymousVoting=$anonymousVoting, category=$category, completedAt=$completedAt, createdAt=$createdAt, createdByApi=$createdByApi, creditCapped=$creditCapped, gameId=$gameId, groupStagesEnabled=$groupStagesEnabled, hideSeeds=$hideSeeds, maxPredictionsPerUser=$maxPredictionsPerUser, notifyUsersWhenMatchesOpen=$notifyUsersWhenMatchesOpen, participantsCount=$participantsCount, predictionMethod=$predictionMethod, predictionsOpenedAt=$predictionsOpenedAt, progressMeter=$progressMeter, quickAdvance=$quickAdvance, requireScoreAgreement=$requireScoreAgreement, startedAt=$startedAt, startedCheckingInAt=$startedCheckingInAt, state=$state, teams=$teams, tieBreaks=$tieBreaks, updatedAt=$updatedAt, descriptionSource=$descriptionSource, fullChallongeUrl=$fullChallongeUrl, liveImageUrl=$liveImageUrl, signUpUrl=$signUpUrl, reviewBeforeFinalizing=$reviewBeforeFinalizing, acceptingPredictions=$acceptingPredictions, participantsLocked=$participantsLocked, gameName=$gameName, participantsSwappable=$participantsSwappable, teamConvertable=$teamConvertable, groupStagesWereStarted=$groupStagesWereStarted, lockedAt=$lockedAt, eventId=$eventId, publicPredictionsBeforeStartTime=$publicPredictionsBeforeStartTime, grandFinalsModifier=$grandFinalsModifier, participants=$participants, matches=$matches)"
    }
}